### PR TITLE
Migrate relevant functionality from test262-parser.

### PR DESCRIPTION
**Pull request by rwaldron**
~~This intentionally omits the changes necessary for hashbang. Once this work is complete, I will call on @adrianheine to apply their changes here, instead of test262-parser.~~

This updated changeset includes @adrianheine's latest work from test262-parser

### CODE REVIEW

File: lib/compiler.js [hunk 1]
```diff
@@ -13,14 +13,7 @@ module.exports = function compile(test, options) {
...
 
-  test = parseFile(test);
-
-  // The following properties are defined by the `test262-parser` module. They
-  // are removed here because they are inconsistent and could therefore cause
-  // confusion.
-  delete test.isATest;
-  delete test.async;
-  delete test.strictMode;
+  const test = new TestFile(descriptor);
```
Comment (jugglinmike) [inline lib/compiler.js]: Awesome!

File: lib/test-file.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
+// Based on
+// https://github.com/smikes/test262-parser/blob/62944f8d000ce7685b917cb7c061301a391a949c/lib/parser.js
+// Copyright (C) 2014, Microsoft Corporation. All rights reserved.
+// This code is governed by the BSD License found in the LICENSE file.
```
Comment (jugglinmike) [inline lib/test-file.js]: The referenced LICENSE file differs from our own. I think we'll need a LICENSE-MICROSOFT file...

Comment (rwaldron) [reply lib/test-file.js]: Fixed!

File: .jshintrc [hunk 3]
```diff
@@ -2,6 +2,5 @@
   "esversion": 6,
   "node": true,
   "undef": true,
-  "unused": true,
-  "validthis": true
+  "unused": true
```
Comment (jugglinmike) [inline .jshintrc]: Nice

File: lib/parser.js [hunk 4]
```diff
@@ -0,0 +1,227 @@
+// Copyright (C) 2014, Microsoft Corporation. All rights reserved.
+// This code is governed by the BSD License found in the LICENSE file.
```
Comment (jugglinmike) [inline lib/parser.js]: LICENSE -> LICENSE-MICROSOFT.txt

Comment (rwaldron) [reply lib/parser.js]: lol, whoops. 

Fixed!

File: .jshintrc (hunk 3 above)
Comment (rwaldron) [reply .jshintrc]: I had to temporarily add `"validthis": true` to allow the "as-is" source to pass, but did so knowing that I could immediately remove it in the follow up changeset.

**Comment by rwaldron** [comment]
@jugglinmike please hold off on merging this, I need to verify the `contents` property contains all of the expected test source code when handling hashbang tests

File: .jshintrc (hunk 3 above)
Comment (jugglinmike) [reply .jshintrc]: That is exactly the nuanced procedure that I was complimenting :)

File: test/test.js [hunk 5]
```diff
@@ -280,3 +281,32 @@ tape('completion detection when final files visited are not tests', t => {
...
+    const hashbang = '#!whatever\n';
+    const expectedContentDir = path.join(fixtureDir, 'expected-content');
+    const expectedMetadataDir = path.join(fixtureDir, 'expected-metadata');
+
+    ids.forEach(id => {
+      const startsWithHashbang = fs.readFileSync(
+        path.join(expectedContentDir, `${id}.js`), 'utf8'
+      ).startsWith(hashbang);
+
+      t.equal(startsWithHashbang, true, `${id} begins with "${hashbang}" (the space was a newline character)`);
```
Comment (jugglinmike) [inline test/test.js]: The files in the `expected-content` directory are checked in to the repository, and since we're not overwriting them during the test, we know they will always include the hashbang. I think we can remove this assertion without losing coverage.

File: test/test.js [hunk 6]
```diff
@@ -280,3 +281,32 @@ tape('completion detection when final files visited are not tests', t => {
...
+
+  stream.on('error', (error) => {
+    t.ok(error);
+    t.end(error);
+  });
+
+  stream.on('end', () => {
+    const hashbang = '#!whatever\n';
+    const expectedContentDir = path.join(fixtureDir, 'expected-content');
+    const expectedMetadataDir = path.join(fixtureDir, 'expected-metadata');
```
Comment (jugglinmike) [inline test/test.js]: This constant is unused, but as noted below, I think much of this code can be removed.

File: test/collateral/valid-with-hashbang/expected-content/test/asyncWithLicense_default.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
+#!whatever
```
Comment (jugglinmike) [inline test/collateral/valid-with-hashbang/expected-content/test/asyncWithLicense_default.js]: A little variation would be good, if only to prove that we haven't hard-coded the value "whatever" into the parser. Though there may be some slightly more interesting edge cases than, say, "whatever2". "  leading space" and "trailing space  " come to mind.

File: lib/normalize-contents.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+
+module.exports = function(contents) {
+  let hashbangOrNot = '';
+
+  if (contents.startsWith('#!')) {
+    let lines = contents.split(/\n/);
```
Comment (jugglinmike) [inline lib/normalize-contents.js]: In the latest version of the proposal, the hashbang is terminated with a LineTerminator--not just a Line Feed. Test262 is also missing tests for other LineTerminators, so [I've submitted a few](https://github.com/tc39/test262/pull/2104)

File: test/collateral/valid-with-hashbang/expected-content/test/asyncWithLicense_default.js (hunk 7 above)
Comment (jugglinmike) [reply test/collateral/valid-with-hashbang/expected-content/test/asyncWithLicense_default.js]: Found more interesting edge cases: the other LineTerminators

**Comment by jugglinmike** [comment]
@rwaldron This made me think that an update to Test262's `INTERPRETING.md` was in order, but I don't think that's true because all of the relevant tests in Test262 have been authored to use the `raw` frontmatter flag.

That's good news for me because I was having a heck of a time writing clear instructions for Test262 consumers. I think it might also mean we can simplify your work here. We don't have to support rebuilding test code to wrangle the hashbang sequence with, e.g. includes and directive prologues. We just have to make sure the metadata is correctly parsed in tests that use a hashbang.

Does that sound right to you?

File: lib/normalize-contents.js (hunk 8 above)
Comment (rwaldron) [reply lib/normalize-contents.js]: Ah! Good catch. I'll work on updating this.

File: test/collateral/valid-with-hashbang/expected-content/test/async-with-ps_strict_mode.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
+#!PS // a U+2029 <PS> PARAGRAPH SEPARATOR is to the immediate left of "//"
```
Comment (rwaldron) [inline test/collateral/valid-with-hashbang/expected-content/test/async-with-ps_strict_mode.js]: @jugglinmike ^^

File: lib/normalize-contents.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
+'use strict';
+
+const LineTerminatorSequence = [
+  '\u000A', // <LF>
+  '\u000D', // <CR>
+  '\u2028', // <LS>
+  '\u2029', // <PS>
+  '\u000D\u000A', // <CR><LS>
+];
```
Comment (rwaldron) [inline lib/normalize-contents.js]: @jugglinmike ^^

**Comment by rwaldron** [comment]
@jugglinmike I've removed the content normalization code and removed the tests that relied on that normalization. I've kept one test that documents the expected behavior—let me know if you want that removed as well.

File: lib/test-file.js [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+ * @private
+ */
+function extractCopyright(test262File) {
+  let lines = test262File.contents.split(/\r|\n/);
+
+  for (let line of lines) {
+    // The very first line may be a hashbang, so look at
+    // all lines until reaching a line that looks close
+    // enough to copyright
+    let result = /^(?:(?:\/\/.*)*)/.exec(line);
```
Comment (jugglinmike) [inline lib/test-file.js]: I missed this RegExp during my first review, but it seems kind of wonky to me. We should revisit it later.

File: lib/test-file.js [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+/**
+ * Extract copyright message
+ *
+ * @method extractCopyright
+ * @param {Test262File} file - file object
+ * @return {string} the copyright string extracted from contents
+ * @private
+ */
+function extractCopyright(test262File) {
+  let lines = test262File.contents.split(/\r|\n/);
```
[the remark this answers is not in the document]
Comment (rwaldron) [reply lib/test-file.js]: I think only if the copyright is on the same line, immediately following the real <PS>, because this works as expected: 

```js
`#!foo 
// Copyright 2001 A Space Odyssey
// There is a real <PS> immediately following '#!foo'`.split(/\r|\n/);
```

Result: 

```
[
  "#!foo ",
  "// Copyright 2001 A Space Odyssey",
  "// There is a real <PS> immediately following '#!foo'",
]
```

However, this would be a problem: 

```js
`#!foo // Copyright 2001 A Space Odyssey
// There is a real <PS> immediately following '#!foo'`.split(/\r|\n/);
```

Result: 

```
[
"#!foo // Copyright 2001 A Space Odyssey",
"// There is a real <PS> immediately following '#!foo'"
]
```

...However... There is presently no restriction in INTERPRETING.md against writing `#!foo // Copyright 2001 A Space Odyssey`. It will also pass lint checks.

**Comment by rwaldron** [comment]
@jugglinmike I've squashed 7 commits into 2. This is ready to go when you are. If we could get a release out today, that would be very cool :D

**Comment by rwaldron** [comment]
Rebased!

**Comment by jugglinmike** [comment]
Nice work, Rick

**Comment by jugglinmike** [comment]
(Tagged as v1.2.1 and published to npm)
